README.md
<h1 align="center">
	<img src="readme_screenshots/Logo.png" alt="Mojave">
</h1>
<h4 align="center">The desktop arcade-style game board for <a href="https://play.battlesnake.com" target="_blank">Battlesnake</a> 🐍</h4>
<p align="center">
  <a href="#about">About</a> •
  <a href="#features">Features</a> •
  <a href="#download-and-install">Download and Install</a> •
  <a href="#managing-snakes">Managing Snakes</a> •
  <a href="#game-setup">Game Setup</a> •
  <a href="#how-to-play">How to Play</a> •
  <a href="#development-tools">Development Tools</a> •
  <a href="#configuration">Configuration</a> •
  <a href="#known-issues">Known Issues</a> •
  <a href="#troubleshooting">Troubleshooting</a> •
  <a href="#credits">Credits</a> •
  <a href="#license">License</a>
</p>

![Gameplay Screenshot](readme_screenshots/gameplay.png?raw=true)

## About
Mojave is a desktop game board and development / debugging tool for [Battlesnake](https://play.battlesnake.com). It aims to bring an immersive "arcade" style experience to the game, with crisp neon visuals, an electrifying soundtrack, and a high amount of customizability.

Don't know Battlesnake? It's an online programming game played by developers from all over the world! Build a snake "bot" and be the last one slithering on the game board to win. You can use any language or technology that implements a web server. For more information on how to write your snake, and on the rules of the game, take a look at the [Battlesnake docs](https://docs.battlesnake.com/).

Battlesnake's mission is to make programming fun and accessible for everyone! With Mojave, we hope to make your Battlesnake experience even better!

## Features
* Customizable visuals for the ultimate sensory experience including a bloom filter, vignette and animations
* Supports all Battlesnake API versions from 2017 and up, simultaneously in one game - try your snake out against the best snakes from years past!
	* Note: Snakes built on API versions earlier than v1 can only play in Standard rules games
* Play with different sets of rules including Royale, Squads, Constrictor, and Wrapped
* Pause, step forward and backward through games to debug your snake's behaviour
* Dump the JSON request/response being sent to/from your snake at any turn
* Includes the [Robosnake](https://github.com/smallsco/robosnake) as a built-in opponent to test your snake against
* Play as a human against your Battlesnake or play solo for that classic Snake experience!

## Download and Install

### Mac OS
For Mac users, the minimum OS requirement is 10.14 (also called Mojave ... hey Apple, I used the name first!)

Apple Silicon users: Mojave has not been tested on the M1. It won't work natively due to the use of shared libraries that will need to be recompiled - but it should run correctly under Rosetta emulation. Let me know if you've tried it!

#### Mac OS Standard Installation
* Download the latest version of Mojave from the [Releases](https://github.com/smallsco/mojave/releases) page and extract the zip file into your Applications folder.
* The first time you run Mojave, you need to **right-click** on the application icon and select the **Open** option from the contextual menu. You will be presented with a warning dialog that the application cannot be opened because the developer cannot be verified. Click **Cancel**.
* Then, repeat the previous step. This time you will have an option to **Open** from the warning dialog. Click **Open** and Mojave will launch. You should only have to do this once per update.

Unfortunately, with each new Mac OS update, Apple makes it increasingly difficult to install and run software that has not been code signed by a member of their developer program. If you still have trouble launching the game despite following the above instructions, you can try running it using LÖVE 11.3 directly via the alternate installation instructions below:

#### Mac OS Alternate Installation
* You will then need to download and install [LÖVE 11.3](https://github.com/love2d/love/releases/tag/11.3) for Mac OS.
	* **DO NOT USE LOVE 11.4 OR GREATER, IT IS NOT COMPATIBLE AND THE GAME WILL CRASH**
* Once you have installed that, you can download the latest version of Mojave from the [Releases](https://github.com/smallsco/mojave/releases) page and extract the zip file somewhere on your computer. **Download the *Linux* version, not the Mac OS version!**
* Open the `Mojave-x.y.z.love` file to start playing!

### Windows
* For Windows users, the minimum requirement is Windows Vista. A 64-bit version of Windows is required.
* You will need to download and install the [Visual C++ 2017 Redistributable](https://go.microsoft.com/fwlink/?LinkId=746572) package from Microsoft if you don't already have it installed on your computer. [Get it here](https://go.microsoft.com/fwlink/?LinkId=746572).
* Once you have installed that, you can download the latest version of Mojave from the [Releases](https://github.com/smallsco/mojave/releases) page and extract the zip file somewhere on your computer.
* Open the `Mojave.exe` application to start playing!

### Linux
* For Linux users, any recent distribution should work.
* First, you will need to install the libcurl development package for your distribution:
    * For Ubuntu/Debian: `sudo apt-get install libcurl4-openssl-dev`
    * For Fedora/Centos: `sudo yum install libcurl-devel`
* You will then need to download and install [LÖVE 11.3](https://github.com/love2d/love/releases/tag/11.3) for your Linux distribution. You can do this via your package manager, or download a package directly from the [LÖVE web site](https://www.love2d.org).
	* **DO NOT USE LOVE 11.4 OR GREATER, IT IS NOT COMPATIBLE AND THE GAME WILL CRASH**
* Once you have installed that, you can download the latest version of Mojave from the [Releases](https://github.com/smallsco/mojave/releases) page and extract the zip file somewhere on your computer.
* Open the `Mojave-x.y.z.love` file to start playing!

### Building from Source
* Clone this repository to somewhere on your computer, or use the latest "Source Code" download from the [Releases](https://github.com/smallsco/mojave/releases) page.
* Navigate to the packager directory: `cd mojave/packager`
* Run the command `./lp-build.sh` from the packager directory.
* This will create a new folder inside the packager directory, `builds`, containing zipped binaries for Windows and Mac OS, and a zipped .love file for Linux.

## Managing Snakes
![Manage Snakes Screen](readme_screenshots/manage_snakes.png?raw=true)
To get started, you'll need to add at least one snake. Click the **Manage Snakes** button on the menu to get to the snake management screen. You will need to select the kind of snake that you would like to add, and press the **Add Snake** button to add it. Mojave supports four kinds of snakes:  

* **Battlesnake API** - Select this option to add a snake that uses the modern "V" Battlesnake API which was launched with the Global Arena. The version will be automatically be detected - snakes using V1 will have a preview rendered.
* **Old Battlesnake API** - Select this option to add a snake that uses the legacy Battlesnake API which changed from year to year and corresponded to a Victoria Live event. You will need to select the API year, 2017 and 2018 are suspported.
* **Robosnake** - The Robosnake is a tough computer-controlled snake built-in to Mojave. Use it as an opponent for testing your snake against! You can add as many as you would like, and will need to select its' color and head/tail types.
* **Human** - Add a Human snake if you would like to play, as a human, against one or more battlesnakes (or even just by yourself for the plain old Snake experience). You can control these snakes by using the arrow keys.

## Game Setup
![Create Game Screen](readme_screenshots/create_game.png?raw=true)
To start a new game, click the **Create Game** button on the menu to get to the game setup screen. You will need to select the rules to play under (see the "How to Play" section below for a description of the different rules), the board size, and which of your snakes you would like to add to this game. If you select the Royale game mode, you will also need to specify how quickly hazards should advance. For both the Royale and Custom game modes, you will need to specify how much damage to take when entering a hazard.

Note that there are a few restrictions on game creation:

* Only snakes using the modern Battlesnake API, or Human snakes may play using the Royale, Squads, Constrictor, Wrapped, and Custom game modes.
* The Royale, Constrictor, Wrapped, and Custom game modes require a minimum of two snakes to play.
* The Squads game mode requires a minimum of two _squads_ to play.
* Only one Human snake may participate in a game.

## How to Play
For full game rules, see the official [Battlesnake Rules](https://docs.battlesnake.com/references/rules), the following is an abridged version:

### Standard
Under Standard rules, snakes may freely roam the game board, and lose some health on each turn. Food will spawn periodically on the game board. Eating food will restore a snake to full health, but will cause it to grow by one tile. A snake will die if it runs out of health, runs into a wall, or into the body of itself or another snake. If two snakes have a head-on-head collision, the smaller snake will be killed (and both will be killed if they are the same size).

### Royale
Under Royale rules, the edges of the board will slowly fill with hazards. If the head of a snake enters a hazard, that snake will take damage at an increased rate.

### Squads
Under Squads rules, all snakes are assigned to one of up to four squads. Snakes that belong to the same squad will share health and length, and may pass through each other's bodies without dying. However, if a snake dies, all other snakes on that squad will die as well.

### Constrictor
Under Constrictor rules, there is no food on the board. Instead, snakes will grow on every turn, with their tail locked to their starting position. This is also known as a Tron game.

### Wrapped
Under Wrapped rules, if a snake crosses the edge of the board, instead of being killed, it will re-enter the board from the opposite edge.

### Custom
Custom games make use of a pre-set template to pre-place food, hazards, and snakes at the start of the game. They are played using Standard rules, however, hazard damage can be customized in a manner similar to Royale games.

## Development Tools
<p align="center">
	<img src="readme_screenshots/development_tools.png">
</p>

You can use the playback controls at the top right of the game window, while a game is taking place, to debug your snake:  
(these controls can also be accessed by using the function keys on your keyboard)

* The "First" button (or F6) will jump to the first turn of the game.
* The "Rewind" button (or F7) will step back one turn.
* The "Play/Pause" button (or F8) will pause and resume game playback (do note that pausing will only pause the UI, the game will continue to run in the background).
* The "Fast Forward" button (or F9) will step forward one turn.
* The "Last" button (or F10) will jump to the last turn of the game.
* The "Rematch" button (or F11) will end the current game and start a new game with the same options.
* The "Close" button (or F12) will end the current game and return to the main menu.


CHANGELOG.md
## CHANGELOG

### v3.5.2 (2022-01-28)

* Fixed visual glitches when playing games in Wrapped mode with the "Curve on Turns" option enabled.

### v3.5.1 (2022-01-23)

* Fixed a regression bug introduced in version 3.4 that would cause starting food to fail to spawn on fixed board sizes.
* Ported changes from the latest version of the official Battlesnake Rules (v1.0.26):
    * Default move is correctly applied when using the Wrapped ruleset.
        * (Mojave already supported this, but we now do it the same way as the official rules).
    * On fixed board sizes, food will only spawn on locations away from the center of the board.

### v3.5 (2021-12-12)

* Substantially rewrote the head and tail loader and renderer in order to support non-monochrome head and tail images.
* Added additional head and tail images.

### v3.4.1 (2021-11-30)

* Added the `customizations` field to the JSON sent to API v1 snakes.

### v3.4 (2021-11-07)

* Custom game modes are now supported! These modes are often featured on Snake Pit Live and include Bridges, Dots, Rings, and Lines.
* Moved hazard damage and speed settings to the game creation screen, displayed for Royale (speed + damage) and Custom (damage only) games.
* Added the "source" parameter to API calls, which will always return "custom" as Mojave doesn't have leagues.
* The heads/tails lists are now sorted alphabetically when adding a Robosnake or human-controlled snake.
* Modified some internal data structures for improved performance and maintainability.
* Renamed "Bloom Filter" option to "Bloom Effect".
* Added additional head and tail images.

### v3.3.1 (2021-10-09)

* You can now use the function keys on your keyboard to control Mojave:
    * F4 - toggles music and SFX
    * F5 - toggles fullscreen mode
    * F6 through F12 - correspond to the in-game controls: first turn, previous turn, pause/resume, next turn, last turn, restart game, and exit.
    * Thank you to [Xtagon](https://github.com/xtagon) for suggesting this feature!

### v3.3 (2021-09-26)

* Mojave can now run at higher resolutions!
    * The game window can be resized, both at the menu and while a game is in progress.
    * The menu background now draws correctly if the game is launched in fullscreen mode.
    * The menu UI scales up correctly if the window is resized or the game enters fullscreen mode.
    * Do note that the minimum required resolution remains 1280x720.
    * There is still UI work to be done to improve _usability_ at higher resolutions, but much of the foundational work is now complete.
* Added the ability to spawn food and hazards deterministically when starting a game, although we're not using it anywhere yet.
* Timeouts used when adding a snake are now hardcoded to 5 seconds rather than using the in-game timeout.
* Added additional head and tail images.

### v3.2.1 (2021-09-04)

* Manual placement of snakes on the game board is now supported, although we don't use this anywhere yet.
* Corrected the text on the Options screen because in some cases, you actually need to press Save Changes for the changes to take effect. Also I had the name of the button wrong...
* Fixed an edge case where hazard damage would default to 15 instead of 14.
* Minor portability fixes.

### v3.2 (2021-09-02)

* Updated Battlesnake Client API:
    * Ruleset configuration parameters are now sent to API v1 snakes in the move JSON.
    * "squad" is now always sent as part of the move JSON.
* Ported changes from latest Battlesnake Rules (v1.0.22):
    * Moved board generation logic out of StandardRules.
    * StandardRules is now responsible for applying hazard damage.
    * Snake health is not reduced when eating food in a hazard.
* Added support for the Wrapped game mode.
* Added a config option to optionally limit the number of turns that a game can run for.
* Reduced the default hazard damage to 14 (Note: if you are upgrading from a previous version, you will need to make this change in the Options yourself)
* Minor UX improvements:
    * Added some text to the Options screen to make it clear that changes to options, while having immediate effect, do not persist unless the save button is pressed.
    * Added some text to the error dialog when trying to add a new snake that says to double-check that the name and URL were entered into the correct fields.
    * Centered the Save / Revert Options buttons to make them more prominent.
    * Left-aligned the text field labels when adding a new snake.
* Snake shouts can now be viewed under the latency when hovering over a snake avatar during a game.
* Disabled some debugging messages that could cause confusion when launching the game for the first time under Linux.
* Added additional head and tail images.

### v3.1.6 (2021-07-19)

* Squad games now send the correct ruleset name in the snake request JSON.
* Fixed a bug that could cause the "shout" key to be dropped for some snakes in the request JSON.

### v3.1.5 (2021-06-12)

* Added additional head and tail images.

### v3.1.4 (2021-04-16)

* Fixed a bug causing hazard deaths in Royale games to not register on the correct turn.
* Corrected tail rendering when being drawn in up/down state.
* Added additional head and tail images.

### v3.1.3 (2021-03-19)

* Fixed a bug that caused the food spawn chance, minimum food, snake max health, and snake start size options to not be respected in non-standard game modes.
* Fixed a bug that would cause the game to crash if a snake response was valid JSON, but was not contained in a dictionary.
* Corrected love-imgui URL to use the correct fork in docs/credits.

### v3.1.2 (2021-03-10)

* Fixed a major bug on Mac OS where Mojave would fail to load ImGui on systems that did not have LuaJIT installed.

modules/Map.lua
local Map = {}
Map.__index = Map
setmetatable( Map, {
  __call = function (cls, ...)
    return cls.new(...)
  end,
})


-- Tile Constants
Map.TILE_FREE = 1       -- Tile contains nothing
Map.TILE_WALL = 2       -- Tile is a wall / blocked
Map.TILE_FOOD = 3       -- Tile contains food
Map.TILE_GOLD = 4       -- Tile contains gold
Map.TILE_SNEK_1 = 11
Map.TILE_SNEK_2 = 12
Map.TILE_SNEK_3 = 13
Map.TILE_SNEK_4 = 14
Map.TILE_SNEK_5 = 15
Map.TILE_SNEK_6 = 16
Map.TILE_SNEK_7 = 17
Map.TILE_SNEK_8 = 18
Map.TILE_SNEK_9 = 19
Map.TILE_SNEK_10 = 20


local function rotatedRectangle( mode, x, y, w, h, rx, ry, segments, r, ox, oy )
	-- Check to see if you want the rectangle to be rounded or not:
	if not oy and rx then r, ox, oy = rx, ry, segments end
	-- Set defaults for rotation, offset x and y
	r = r or 0
	ox = ox or w / 2
	oy = oy or h / 2
	love.graphics.push()
    love.graphics.translate( x + ox, y + oy )
    love.graphics.rotate( -r )
    love.graphics.rectangle( mode, -ox, -oy, w, h, rx, ry, segments )
	love.graphics.pop()
end


--- Constructor / Factory Function
-- @param table opt A table containing initialization options
-- @return Map
function Map.new( opt )
    
    local self = setmetatable( {}, Map )
    local opt = opt or {}
    
    -- screenWidth and screenHeight: Size of the window
    -- pixelWidth and pixelHeight: Size of the board
    local screenWidth, screenHeight = love.graphics.getDimensions()
    self.pixelWidth = screenWidth * 0.8
    self.pixelHeight = screenHeight * 0.8
    
    self.bloom = love.graphics.newCanvas( self.pixelWidth/4, self.pixelHeight/4 )
    self.hblur = love.graphics.newCanvas( self.pixelWidth/4, self.pixelHeight/4 )
    self.vblur = love.graphics.newCanvas( self.pixelWidth/4, self.pixelHeight/4 )
    self.bloomscene = love.graphics.newCanvas( self.pixelWidth, self.pixelHeight )
    self.scene = love.graphics.newCanvas( self.pixelWidth, self.pixelHeight )
    
    self.vxScale = self.pixelWidth / bgVignette:getWidth()
    self.vyScale = self.pixelHeight / bgVignette:getHeight()
    
    -- How many tiles/squares to fit into pixelWidth/pixelHeight
    self.numTilesX = opt.width or 25
    self.numTilesY = opt.height or 15
    
    -- Compute the width and height of each tile
    self.tileWidth = self.pixelWidth / self.numTilesX
    self.tileHeight = self.pixelHeight / self.numTilesY
    
    -- The "cell" is the part of each tile that we will actually draw
    -- the rest of the tile is reserved to allow the background to show through
    local outlineSize = opt.outlineSize or 1
    self.cellWidth = self.tileWidth - outlineSize
    self.cellHeight = self.tileHeight - outlineSize
    
    -- Generate the tile grid
    self.tiles = {}
    self:clear()
    
    return self
    
end

function Map:clear()

    for i = 1, self.numTilesY do
        self.tiles[i] = {}
        for j = 1, self.numTilesX do
            self.tiles[i][j] = Map.TILE_FREE
        end
    end

end

function Map:print()
    
    local chr = ''
    local str = ''
    for y = 1, self.numTilesY do
        for x = 1, self.numTilesX do
            if self.tiles[y][x] == Map.TILE_FREE then
                chr = '.'
            elseif self.tiles[y][x] == Map.TILE_WALL then
                chr = 'X'
            elseif self.tiles[y][x] == Map.TILE_GOLD then
                chr = 'G'
            elseif self.tiles[y][x] == Map.TILE_FOOD then
                chr = 'F'
            elseif self.tiles[y][x] == Map.TILE_SNEK_1 then
                chr = '1'
            elseif self.tiles[y][x] == Map.TILE_SNEK_2 then
                chr = '2'
            elseif self.tiles[y][x] == Map.TILE_SNEK_3 then
                chr = '3'
            elseif self.tiles[y][x] == Map.TILE_SNEK_4 then
                chr = '4'
            elseif self.tiles[y][x] == Map.TILE_SNEK_5 then
                chr = '5'
            elseif self.tiles[y][x] == Map.TILE_SNEK_6 then
                chr = '6'
            elseif self.tiles[y][x] == Map.TILE_SNEK_7 then
                chr = '7'
            elseif self.tiles[y][x] == Map.TILE_SNEK_8 then
                chr = '8'
            elseif self.tiles[y][x] == Map.TILE_SNEK_9 then
                chr = '9'
            elseif self.tiles[y][x] == Map.TILE_SNEK_10 then
                chr = '0'
            end
            str = str .. '[' .. chr .. ']'
        end
        str = str .. "\n"
    end
    gameLog( str, 'info' )
    
end

function Map:draw( mySnakes, food, gold, walls )

    -- clear canvases
    love.graphics.setCanvas( self.bloom, self.hblur, self.vblur )
	love.graphics.clear()
	love.graphics.setCanvas( self.bloomscene, self.scene )
	love.graphics.clear()
	love.graphics.setCanvas()
    love.graphics.clear()
	
	-- base, quarter scale base without bg
	love.graphics.setCanvas( self.scene )
	self:draw2( true, mySnakes, food, gold, walls )
	love.graphics.setCanvas( self.bloomscene )
    self:draw2( false, mySnakes, food, gold, walls )
    love.graphics.setColor( 255, 255, 255, 255 )
    local blendmode, blendalphamode = love.graphics.getBlendMode()
    love.graphics.setBlendMode( "alpha", "premultiplied" )
    
    -- apply bloom effect
    love.graphics.push()
    love.graphics.scale( 0.25, 0.25 )
    love.graphics.setCanvas( self.bloom )
    love.graphics.setShader( Shaders.bloom )
    love.graphics.draw( self.bloomscene )
    love.graphics.pop()
    
    -- apply horizontal blur
    love.graphics.setCanvas( self.hblur )
    love.graphics.setShader( Shaders.horizontalblur )
    love.graphics.draw( self.bloom )
    
    -- apply vertical blur
    love.graphics.setCanvas( self.vblur )
    love.graphics.setShader( Shaders.verticalblur )
    love.graphics.draw( self.hblur )
    
    -- final scene
    love.graphics.setCanvas()
    Shaders.combine:send( "bloomtex", self.vblur )
    love.graphics.setShader( Shaders.combine )
    love.graphics.draw( self.scene )
    love.graphics.setShader()
    love.graphics.setBlendMode( blendmode, blendalphamode )

end

--- Draws the map to the screen.
function Map:draw2( drawgrid, mySnakes, food, gold, walls )
    
    if drawgrid then
        -- Draw the grid
        for i = 1, self.numTilesY do
            for j = 1, self.numTilesX do
                if i % 2 == 0 and j % 2 == 0 then
                    love.graphics.setColor(
                        Util.denormalizeRGBArray(
                            config[ 'appearance' ][ 'tileSecondaryColor' ]
                        )
                    )
                else
                    love.graphics.setColor(
                        Util.denormalizeRGBArray(
                            config[ 'appearance' ][ 'tilePrimaryColor' ]
                        )
                    )
                end
                love.graphics.rectangle(
                    "fill",
                    1 + ((j-1) * self.tileWidth) + ((self.tileWidth - self.cellWidth) / 2),
                    1 + ((i-1) * self.tileHeight) + ((self.tileHeight - self.cellHeight) / 2),
                    self.cellWidth,
                    self.cellHeight
                )
            end
        end
        
        -- vignette
        if config[ 'appearance' ][ 'enableVignette' ] then
            love.graphics.setColor( 255, 255, 255, 255 )
            love.graphics.draw( bgVignette, 0, 0, 0, self.vxScale, self.vyScale )
        end
    end
    
    -- Walls
    love.graphics.setColor(
        Util.denormalizeRGBArray(
            config[ 'appearance' ][ 'wallColor' ]
        )
    )
    for i = 1, #walls do
        love.graphics.rectangle(
            "fill",
            1 + ((walls[i][1]-1) * self.tileWidth) + ((self.tileWidth - self.cellWidth) / 2),
            1 + ((walls[i][2]-1) * self.tileHeight) + ((self.tileHeight - self.cellHeight) / 2),
            self.cellWidth,
            self.cellHeight
        )
    end
    
    -- Food
    love.graphics.setColor(
        Util.denormalizeRGBArray(
            config[ 'appearance' ][ 'foodColor' ]
        )
    )
    for i = 1, #food do
        rotatedRectangle(
            "fill",
            1 + ((food[i][1]-1) * self.tileWidth) + ((self.tileWidth - self.cellWidth) / 2),
            1 + ((food[i][2]-1) * self.tileHeight) + ((self.tileHeight - self.cellHeight) / 2),
            self.cellWidth,
            self.cellHeight,
            nil,
            nil,
            nil,
            math.fmod(love.timer.getTime(), 2 * math.pi)
        )
    end
    
    -- Gold
    love.graphics.setColor(
        Util.denormalizeRGBArray(
            config[ 'appearance' ][ 'goldColor' ]
        )
    )
    for i = 1, #gold do
        rotatedRectangle(
            "fill",
            1 + ((gold[i][1]-1) * self.tileWidth) + ((self.tileWidth - self.cellWidth) / 2),
            1 + ((gold[i][2]-1) * self.tileHeight) + ((self.tileHeight - self.cellHeight) / 2),
            self.cellWidth,
            self.cellHeight,
            nil,
            nil,
            nil,
            math.fmod(love.timer.getTime(), 2 * math.pi)
        )
    end
    
    -- Warps?
    -- Reverse?
    --[[love.graphics.setColor(255,96,222,234)
    rotatedRectangle(
        "fill",
        1 + ((9-1) * self.tileWidth) + ((self.tileWidth - self.cellWidth) / 2),
        1 + ((5-1) * self.tileHeight) + ((self.tileHeight - self.cellHeight) / 2),
        self.cellWidth,
        self.cellHeight,
        nil,
        nil,
        nil,
        math.fmod(love.timer.getTime(), 2 * math.pi)
    )]]
    
    -- Sneks
    for i = 1, #mySnakes do
        for j = 1, #mySnakes[i][ 'position' ] do
        
            -- dead snakes are completely translucent
            -- living snakes get slightly translucent from their head to their tail
            local alpha = 32
            if mySnakes[i][ 'alive' ] then
                if config[ 'appearance' ][ 'fadeOutTails' ] then
                    alpha = ( ( ( #mySnakes[i][ 'position' ] - j ) * 150 ) / #mySnakes[i][ 'position' ] ) + 105
                else
                    alpha = 255
                end
            end
            love.graphics.setColor( mySnakes[i][ 'color' ][1], mySnakes[i][ 'color' ][2], mySnakes[i][ 'color' ][3], alpha )
            
            if j ~= 1 and j ~= #mySnakes[i][ 'position' ] then
            
                if
                    mySnakes[i][ 'position' ][j][1] == mySnakes[i][ 'position' ][ #mySnakes[i][ 'position' ] ][1]
                    and mySnakes[i][ 'position' ][j][2] == mySnakes[i][ 'position' ][ #mySnakes[i][ 'position' ] ][2]
                then
                    -- noop
                else
                    -- body
                    love.graphics.rectangle(
                        "fill",
                        1 + ((mySnakes[i][ 'position' ][j][1]-1) * self.tileWidth) + ((self.tileWidth - self.cellWidth) / 2),
                        1 + ((mySnakes[i][ 'position' ][j][2]-1) * self.tileHeight) + ((self.tileHeight - self.cellHeight) / 2),
                        self.cellWidth,
                        self.cellHeight
                    )
                end
            
            elseif j == #mySnakes[i][ 'position' ] and j ~= 1 then
            
                if
                    mySnakes[i][ 'position' ][j][1] == mySnakes[i][ 'position' ][1][1]
                    and mySnakes[i][ 'position' ][j][2] == mySnakes[i][ 'position' ][1][2]
                then
                    -- noop
                else
                    -- tail
                    local r = 0
                    local xs = self.cellWidth / mySnakes[i][ 'tail' ]:getWidth()
                    local ys = self.cellHeight / mySnakes[i][ 'tail' ]:getHeight()
                    if mySnakes[i][ 'position' ][j][3] == Snake.DIRECTION_NORTH then
                        r = math.rad(270)
                        local xs2 = ys
                        ys = xs
                        xs = xs2
                    elseif mySnakes[i][ 'position' ][j][3] == Snake.DIRECTION_SOUTH then
                        r = math.rad(90)
                        local xs2 = ys
                        ys = xs
                        xs = xs2
                    elseif mySnakes[i][ 'position' ][j][3] == Snake.DIRECTION_WEST then
                        xs = -xs
                    end
                    love.graphics.draw(
                        mySnakes[i][ 'tail' ],
                        1 + ((mySnakes[i][ 'position' ][j][1]-1) * self.tileWidth) + ((self.tileWidth - self.cellWidth) / 2) + ( self.cellWidth / 2 ),
                        1 + ((mySnakes[i][ 'position' ][j][2]-1) * self.tileHeight) + ((self.tileHeight - self.cellHeight) / 2) + ( self.cellHeight / 2 ),
                        r,
                        xs,
                        ys,
                        mySnakes[i][ 'tail' ]:getWidth() / 2,
                        mySnakes[i][ 'tail' ]:getHeight() / 2
                    )
                end
            
            elseif j == 1 then
            
                -- head
                if
                    mySnakes[i][ 'position' ][j][1] < 1 or
                    mySnakes[i][ 'position' ][j][2] < 1 or
                    mySnakes[i][ 'position' ][j][1] > config[ 'gameplay' ][ 'boardWidth' ] or
                    mySnakes[i][ 'position' ][j][2] > config[ 'gameplay' ][ 'boardHeight' ]
                then
                    -- This is a dead (or about to die) snake that went off the edge
                    -- of the world, so don't try and draw its' head.
                else
                    local r = 0
                    local xs = self.cellWidth / mySnakes[i][ 'head' ]:getWidth()
                    local ys = self.cellHeight / mySnakes[i][ 'head' ]:getHeight()
                    if mySnakes[i][ 'position' ][j][3] == Snake.DIRECTION_NORTH then
                        r = math.rad(270)
                        local xs2 = ys
                        ys = xs
                        xs = xs2
                    elseif mySnakes[i][ 'position' ][j][3] == Snake.DIRECTION_SOUTH then
                        r = math.rad(90)
                        local xs2 = ys
                        ys = xs
                        xs = xs2
                    elseif mySnakes[i][ 'position' ][j][3] == Snake.DIRECTION_WEST then
                        xs = -xs
                    end
                    love.graphics.draw(
                        mySnakes[i][ 'head' ],
                        1 + ((mySnakes[i][ 'position' ][j][1]-1) * self.tileWidth) + ((self.tileWidth - self.cellWidth) / 2) + ( self.cellWidth / 2 ),
                        1 + ((mySnakes[i][ 'position' ][j][2]-1) * self.tileHeight) + ((self.tileHeight - self.cellHeight) / 2) + ( self.cellHeight / 2 ),
                        r,
                        xs,
                        ys,
                        mySnakes[i][ 'head' ]:getWidth() / 2,
                        mySnakes[i][ 'head' ]:getHeight() / 2
                    )
                end

            end
        end
    end
	
	-- reset brush color
	love.graphics.setColor( 255, 255, 255, 255 )
    
end

--- Returns the value of a tile
-- @param x The X coordinate of the game board
-- @param y The Y coordinate of the game board
-- @return The value of the tile at the given X and Y coordinates
function Map:getTile( x, y )
    return self.tiles[y][x]
end

--- Sets the value of a tile
-- @param x The X coordinate of the game board
-- @param y The Y coordinate of the game board
-- @param value The value to set the tile to
function Map:setTile( x, y, value )
    self.tiles[y][x] = value
end

--- Sets a tile of a specific type at a free location near the center of the map
-- @param value The value to set the tile to
-- @return The x and y coordinates that were selected
-- @see http://stackoverflow.com/a/398532/2578262
function Map:setTileAtFreeLocationNearCenter( value )    
    local cx = math.ceil( self.numTilesX / 2 )
    local cy = math.ceil( self.numTilesY / 2 )
    local direction = 1
    local distance = 1
    
    if self.tiles[cy][cx] == Map.TILE_FREE then
        self.tiles[cy][cx] = value
        return cx, cy
    end
    
    while ( math.abs(cx) <= self.numTilesX or math.abs(cy) <= self.numTilesY ) do
    
        for i = 1, distance do
            cx = cx + direction
            if ( math.abs(cx) <= self.numTilesX and math.abs(cy) <= self.numTilesY ) then
                if self.tiles[cy][cx] == Map.TILE_FREE then
                    self.tiles[cy][cx] = value
                    return cx, cy
                end
            end
        end
        
        for i = 1, distance do
            cy = cy + direction
            if ( math.abs(cx) <= self.numTilesX and math.abs(cy) <= self.numTilesY ) then
                if self.tiles[cy][cx] == Map.TILE_FREE then
                    self.tiles[cy][cx] = value
                    return cx, cy
                end
            end
        end
        
        distance = distance + 1
        direction = direction * -1
    end
    
    error( 'No free spaces available on the game board' )
         
end

--- Sets a tile of a specific type at a random, free location
-- @param value The value to set the tile to
-- @param border A border of tiles around the edge of the map that cannot be selected
-- @param evenOnly If true, forces the location to be even-numbered
-- @return The x and y coordinates that were selected
function Map:setTileAtRandomFreeLocation( value, border, evenOnly )
    local x, y
    repeat
        if border then
            x = love.math.random( border, self.numTilesX - border + 1 )
            y = love.math.random( border, self.numTilesY - border + 1 )
        else
            x = love.math.random( self.numTilesX )
            y = love.math.random( self.numTilesY )
        end
    until self.tiles[y][x] == Map.TILE_FREE and ( ( not evenOnly ) or ( evenOnly and ( y + x ) % 2 == 0 ) )
    
    self.tiles[y][x] = value
    return x, y
end

--- Sets a tile of a specific type at a random, safe location
-- @param value The value to set the tile to
-- @oaram badCoords A table of coordinates where this tile cannot be placed
-- @return The x and y coordinates that were selected
function Map:setTileAtRandomSafeLocation( value, badCoords )
    
    local x, y, not_bad_coord, tile_is_free
    repeat
        not_bad_coord = true
        tile_is_free = true
        x = love.math.random( self.numTilesX )
        y = love.math.random( self.numTilesY )
        for i = 1, #badCoords do
            if badCoords[i][1] == x and badCoords[i][2] == y then
                not_bad_coord = false
            end
        end
        if self.tiles[y][x] ~= Map.TILE_FREE then
            tile_is_free = false
        end
    until not_bad_coord == true and tile_is_free == true
    
    self.tiles[y][x] = value
    return x, y
end

return Map